src/librustc/ty/sty.rs

```diff
@@ -738,8 +733,16 @@ impl<'a, 'tcx, 'gcx> PolyExistentialProjection<'tcx> {
         self.0.item_name // safe to skip the binder to access a name
     }
 
-    pub fn sort_key(&self) -> (DefId, Name) {
-        (self.0.trait_ref.def_id, self.0.item_name)
+    pub fn sort_key(&self, hasher: &mut DefPathHashes) -> (u64, InternedString) {
+        // We want something here that is stable across trait boundaries.
```

Comment:
*across crate boundaries. Shouldn't you store the hash on the `TraitDef` rather than recomputing it on every `mk_trait`?